backend/tracim_backend/lib/core/user.py

```diff
@@ -1230,3 +1244,117 @@ def get_about_user(self, user_id: int) -> AboutUser:
             authored_content_revisions_count=content_revisions_infos.count,
             authored_content_revisions_space_count=content_revisions_infos.space_count,
         )
+
+    def get_avatar(
+        self, user_id: int, filename: str, default_filename: str, force_download: bool = False,
+    ) -> HapicFile:
+        user = self.get_one(user_id)
+        if not user.avatar:
+            raise UserAvatarNotFound("avatar of user {} not found".format(user_id))
+        return StorageLib(self._config).get_raw_file(
+            depot_file=user.avatar,
+            filename=filename,
+            default_filename=default_filename,
+            force_download=force_download,
+        )
+
+    def get_avatar_preview(
+        self,
+        user_id: int,
+        filename: str,
+        default_filename: str,
+        width: int = None,
+        height: int = None,
+        force_download: bool = False,
+    ) -> HapicFile:
+        user = self.get_one(user_id)
+        if not user.cropped_avatar:
+            raise UserAvatarNotFound("cropped version of user {} avatar not found".format(user_id))
+        _, original_file_extension = os.path.splitext(self._user.cropped_avatar.filename)
+        return StorageLib(self._config).get_jpeg_preview(
+            depot_file=user.cropped_avatar,
+            filename=filename,
+            default_filename=default_filename,
+            width=width,
+            height=height,
+            original_file_extension=original_file_extension,
+            force_download=force_download,
+            page_number=1,
+        )
+
+    def set_avatar(
+        self, user_id: int, new_filename: str, new_mimetype: str, new_content: typing.BinaryIO
+    ) -> None:
+        user = self.get_one(user_id)
+        label, file_extension = os.path.splitext(new_filename)
+        with tempfile.SpooledTemporaryFile(mode="wb+") as tmp_file:
+            tmp_file.write(new_content.read())
+            tmp_file.seek(0, 0)
+            user.avatar = FileIntent(tmp_file.read(), new_filename, new_mimetype)
+            tmp_file.seek(0, 0)
+            with tempfile.SpooledTemporaryFile(mode="wb+") as cropped_tmp_file:
+                # FIXME - G.M - 2021-01-21 - should we catch error here,
+                # what happened if pillow failed ?
+                crop_image(tmp_file, cropped_tmp_file, ratio=AVATAR_RATIO)
```

Comment:
Why did you choose to have a temporary file? Wasn't it possible tocreate an image from the read contents directly in memory (and crop it)?

Reply:
I do need a copy of the original stream as this is a stream that i can reread and i do need a fileobj to store there the cropped image result to send to FileIntent.
I do agree that other kind of fileobj can be used instead, i'm not sure what is the best solution here.
note that as far i understood SpooledTemporaryFile it seems to me "in memory until reach this size" which seems to me a nice compromise.